src/recommendations/src/recommendations-service/app.py

```diff
@@ -672,6 +676,159 @@ def choose_discounted():
+                chosen_offer_id = offer_ids[int(user_id) % len(offer_ids)]
+                chosen_score = None
+            else:
+                resp_headers['X-Personalize-Recipe'] = get_recipe(campaign_arn)
+                logger.info(f"Input to Personalized Ranking for offers: userId: {user_id}({type(user_id)}) "
+                            f"inputList: {offer_ids}")
+                get_recommendations_response = personalize_runtime.get_recommendations(
+                    campaignArn=campaign_arn,
+                    userId=user_id,
+                    numResults=len(offer_ids)
+                )
+
+                logger.info(f'Recommendations returned: {json.dumps(get_recommendations_response)}')
+
+                return_key = 'itemList'
+                # Here is where might want to incorporate some business logic
+                # for more information on how these scores are used see
+                # https://aws.amazon.com/blogs/machine-learning/introducing-recommendation-scores-in-amazon-personalize/
+                user_scores = {item['itemId']: float(item['score']) for item in
+                               get_recommendations_response[return_key]}
+                # We assume we have pre-calculated the adjusting factor, can be a mix of probability when applicable,
+                # calculation of expected return per offer, etc.
+                adjusted_scores = {offer_id: score * offers_by_id[offer_id]['preference']
+                                   for offer_id, score in user_scores.items()}
+                logger.info(f"Scores after adjusting for preference parameters: {adjusted_scores}")
+
+                # Normalise these - makes it easier to do further adjustments
+                score_sum = sum(adjusted_scores.values())
+                adjusted_scores = {offer_id: score / score_sum
+                                   for offer_id, score in adjusted_scores.items()}
+                logger.info(f"Scores after normalising adjusted scores: {adjusted_scores}")
+
+                # Just one way we could add some randomness - adds serendipity though removes personalization a bit
+                # because we have the scores though we retain a lot of the personalization
+                random_factor = 0.0
+                adjusted_scores = {offer_id: score*(1-random_factor) + random_factor*random.random()
+                                   for offer_id, score in adjusted_scores.items()}
+                logger.info(f"Scores after adding randomness: {adjusted_scores}")
+
+                # We can do many other things here, like randomisation, normalisation in different dimensions,
+                # tracking per-user, offer, quotas, etc. Here, we just select the most promising adjusted score
+                chosen_offer_id = max(adjusted_scores, key=adjusted_scores.get)
+                chosen_score = user_scores[chosen_offer_id]
+                chosen_adjusted_score = adjusted_scores[chosen_offer_id]
```

Comment:
Oh yes - it would be great to show off new this feature ... but I feel it is getting out of scope of the current work. In the existing solution the offers notebook does explore this issue a bit and can be a starting point for customers to think about what they want to do. We can for now put some more information in the notebook and code about this option.